### ISSUE TITLE
TypeError while running SAM algorithm

### ISSUE THREAD

**Issue by alkhairohr** [scverse/scanpy#1293]
Hello,

I am trying to use the SAM algorithm in my single-cell analysis. I run the SAM function like so:

```py
sam_obj = sce.tl.sam(adata,inplace=True)
```

The function runs fine and appears to finish training however it crashes when computing the UMAP with the following error:

```pytb
TypeError: a bytes-like object is required, not 'list'
```

I'm not sure where this problem is coming from and I have spent the past day installing different versions of python and other dependencies to see if that solves the issue. Maybe naive but I know conda can sometimes be behind in their updates.

I installed scanpy following the anaconda instructions here: https://scanpy.readthedocs.io/en/stable/installation.html
And I installed sam-algorithm using pip.

Below is the entire output from the function call above. Below this I have included the output of "conda list" in case this information is helpful. 

Any help would be greatly appreciated. Thank you, Hasan.

```pytb
Self-assembling manifold
Running SAM
RUNNING SAM
Iteration: 0, Convergence: 0.6008695832027542
/wynton/home/state/alkhairohr/miniconda3/envs/python_env/lib/python3.6/site-packages/samalg/__init__.py:1165: FutureWarning: This location for 'connectivities' is deprecated. It has been moved to .obsp[connectivities], and will not be accesible here in a future version of anndata.
  self.adata.uns["neighbors"]["connectivities"] = EDM
Iteration: 1, Convergence: 0.3743130193917588
/wynton/home/state/alkhairohr/miniconda3/envs/python_env/lib/python3.6/site-packages/samalg/__init__.py:1165: FutureWarning: This location for 'connectivities' is deprecated. It has been moved to .obsp[connectivities], and will not be accesible here in a future version of anndata.
  self.adata.uns["neighbors"]["connectivities"] = EDM
Iteration: 2, Convergence: 0.029142717058066172
/wynton/home/state/alkhairohr/miniconda3/envs/python_env/lib/python3.6/site-packages/samalg/__init__.py:1165: FutureWarning: This location for 'connectivities' is deprecated. It has been moved to .obsp[connectivities], and will not be accesible here in a future version of anndata.
  self.adata.uns["neighbors"]["connectivities"] = EDM
Iteration: 3, Convergence: 0.007035797302954891
/wynton/home/state/alkhairohr/miniconda3/envs/python_env/lib/python3.6/site-packages/samalg/__init__.py:1165: FutureWarning: This location for 'connectivities' is deprecated. It has been moved to .obsp[connectivities], and will not be accesible here in a future version of anndata.
  self.adata.uns["neighbors"]["connectivities"] = EDM
Computing the UMAP embedding...
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-10e75f5bfbdb> in <module>
----> 1 sam_obj = sce.tl.sam(adata,inplace=True)

~/miniconda3/envs/python_env/lib/python3.6/site-packages/scanpy/external/tl/_sam.py in sam(adata, max_iter, num_norm_avg, k, distance, standardization, weight_pcs, sparse_pca, n_pcs, n_genes, projection, inplace, verbose)
    216         projection=projection,
    217         sparse_pca=sparse_pca,
--> 218         verbose=verbose,
    219     )
    220 

~/miniconda3/envs/python_env/lib/python3.6/site-packages/samalg/__init__.py in run(self, max_iter, verbose, projection, stopping_condition, num_norm_avg, k, distance, preprocessing, npcs, n_genes, weight_PCs, sparse_pca, proj_kwargs, project_weighted, weight_mode)
   1061             if verbose:
   1062                 print("Computing the UMAP embedding...")
-> 1063             self.run_umap(**proj_kwargs)
   1064         elif projection == "diff_umap":
   1065             if verbose:

~/miniconda3/envs/python_env/lib/python3.6/site-packages/samalg/__init__.py in run_umap(self, X, metric, **kwargs)
   1223             # print(X.shape)
   1224             umap_obj = umap.UMAP(metric=metric, **kwargs)
-> 1225             umap2d = umap_obj.fit_transform(X)
   1226             self.adata.obsm["X_umap"] = umap2d
   1227             self.umap_obj = umap_obj

~/miniconda3/envs/python_env/lib/python3.6/site-packages/umap/umap_.py in fit_transform(self, X, y)
   2010             Embedding of the training data in low-dimensional space.
   2011         """
-> 2012         self.fit(X, y)
   2013         return self.embedding_
   2014 

~/miniconda3/envs/python_env/lib/python3.6/site-packages/umap/umap_.py in fit(self, X, y)
   1831                 self._search_graph.data = _data
   1832                 self._search_graph = self._search_graph.maximum(
-> 1833                     self._search_graph.transpose()
   1834                 ).tocsr()
   1835 

~/miniconda3/envs/python_env/lib/python3.6/site-packages/scipy/sparse/lil.py in transpose(self, axes, copy)
    435 
    436     def transpose(self, axes=None, copy=False):
--> 437         return self.tocsr(copy=copy).transpose(axes=axes, copy=False).tolil(copy=False)
    438 
    439     transpose.__doc__ = spmatrix.transpose.__doc__

~/miniconda3/envs/python_env/lib/python3.6/site-packages/scipy/sparse/lil.py in tocsr(self, copy)
    460             indptr = np.empty(M + 1, dtype=idx_dtype)
    461             indptr[0] = 0
--> 462             _csparsetools.lil_get_lengths(self.rows, indptr[1:])
    463             np.cumsum(indptr, out=indptr)
    464             nnz = indptr[-1]

_csparsetools.pyx in scipy.sparse._csparsetools.lil_get_lengths()

~/miniconda3/envs/python_env/lib/python3.6/site-packages/scipy/sparse/_csparsetools.cpython-36m-x86_64-linux-gnu.so in View.MemoryView.memoryview_cwrapper()

~/miniconda3/envs/python_env/lib/python3.6/site-packages/scipy/sparse/_csparsetools.cpython-36m-x86_64-linux-gnu.so in View.MemoryView.memoryview.__cinit__()

TypeError: a bytes-like object is required, not 'list'
```

Conda list output

```
# packages in environment at /wynton/home/state/alkhairohr/miniconda3/envs/python_env:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
anndata                   0.7.3                    pypi_0    pypi
attrs                     19.3.0                     py_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
blas                      1.0                         mkl  
bleach                    3.1.5              pyh9f0ad1d_0    conda-forge
blosc                     1.19.0               hd408876_0  
bzip2                     1.0.8                h7b6447c_0  
ca-certificates           2020.1.1                      0  
certifi                   2020.6.20                py36_0  
colorlover                0.3.0                      py_0    conda-forge
cycler                    0.10.0                   py36_0  
dbus                      1.13.6               he372182_0    conda-forge
decorator                 4.4.2                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
entrypoints               0.3             py36h9f0ad1d_1001    conda-forge
expat                     2.2.9                he1b5a44_2    conda-forge
fontconfig                2.13.1            he4413a7_1000    conda-forge
freetype                  2.10.2               he06d7ca_0    conda-forge
get-version               2.1                      pypi_0    pypi
glib                      2.63.1               h3eb4bd4_1  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb31296c_0  
h5py                      2.10.0                   pypi_0    pypi
hdf5                      1.10.4               hb1b8bf9_0  
icu                       58.2              hf484d3e_1000    conda-forge
importlib-metadata        1.6.1            py36h9f0ad1d_0    conda-forge
importlib_metadata        1.6.1                         0    conda-forge
intel-openmp              2020.1                      217  
ipyevents                 0.7.1                      py_0    conda-forge
ipykernel                 5.3.0            py36h95af2a2_0    conda-forge
ipython                   7.15.0           py36h9f0ad1d_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.5.1                      py_0    conda-forge
jedi                      0.17.1           py36h9f0ad1d_0    conda-forge
jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
joblib                    0.15.1                     py_0  
jpeg                      9d                   h516909a_0    conda-forge
jsonschema                3.2.0            py36h9f0ad1d_1    conda-forge
jupyter                   1.0.0                      py_2    conda-forge
jupyter_client            6.1.3                      py_0    conda-forge
jupyter_console           6.1.0                      py_1    conda-forge
jupyter_core              4.6.3            py36h9f0ad1d_1    conda-forge
kiwisolver                1.2.0            py36hfd86e86_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
legacy-api-wrap           1.2                      pypi_0    pypi
leidenalg                 0.8.0                    pypi_0    pypi
libedit                   3.1.20191231         h7b6447c_0  
libffi                    3.3                  he6710b0_1  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libpng                    1.6.37               hed695b0_1    conda-forge
libsodium                 1.0.17               h516909a_0    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
libuuid                   2.32.1            h14c3975_1000    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxml2                   2.9.10               he19cac6_1  
llvmlite                  0.33.0                   pypi_0    pypi
lz4-c                     1.9.2                he6710b0_0  
lzo                       2.10                 h7b6447c_2  
markupsafe                1.1.1            py36h8c4c3a4_1    conda-forge
matplotlib                3.2.2                         0  
matplotlib-base           3.2.2            py36hef1b27d_0  
mistune                   0.8.4           py36h8c4c3a4_1001    conda-forge
mkl                       2020.1                      217  
mkl-service               2.3.0            py36he904b0f_0  
mkl_fft                   1.1.0            py36h23d657b_0  
mkl_random                1.1.1            py36h0573a6f_0  
mock                      4.0.2                      py_0  
natsort                   7.0.1                    pypi_0    pypi
nbconvert                 5.6.1            py36h9f0ad1d_1    conda-forge
nbformat                  5.0.6                      py_0    conda-forge
ncurses                   6.2                  he6710b0_1  
networkx                  2.4                      pypi_0    pypi
notebook                  6.0.3            py36h9f0ad1d_0    conda-forge
numba                     0.50.0                   pypi_0    pypi
numexpr                   2.7.1            py36h423224d_0  
numpy                     1.19.0                   pypi_0    pypi
numpy-base                1.18.5           py36hde5b4d6_0  
openssl                   1.1.1g               h7b6447c_0  
packaging                 20.4               pyh9f0ad1d_0    conda-forge
pandas                    1.0.5            py36h0573a6f_0  
pandoc                    2.9.2.1                       0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parso                     0.7.0              pyh9f0ad1d_0    conda-forge
patsy                     0.5.1                    py36_0  
pcre                      8.44                 he1b5a44_0    conda-forge
pexpect                   4.8.0            py36h9f0ad1d_1    conda-forge
pickleshare               0.7.5           py36h9f0ad1d_1001    conda-forge
pip                       20.1.1                   py36_1  
plotly                    4.0.0                      py_0    plotly
prometheus_client         0.8.0              pyh9f0ad1d_0    conda-forge
prompt-toolkit            3.0.5                      py_1    conda-forge
prompt_toolkit            3.0.5                         1    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
pygments                  2.6.1                      py_0    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyqt                      5.9.2            py36hcca6a23_4    conda-forge
pyrsistent                0.16.0           py36h8c4c3a4_0    conda-forge
pytables                  3.6.1            py36h71ec239_0  
python                    3.6.10               h7579374_2  
python-dateutil           2.8.1                      py_0    conda-forge
python-igraph             0.8.2                    pypi_0    pypi
python_abi                3.6                     1_cp36m    conda-forge
pytz                      2020.1                     py_0  
pyzmq                     19.0.1           py36h9947dbf_0    conda-forge
qt                        5.9.7                h5867ecd_1  
qtconsole                 4.7.5              pyh9f0ad1d_0    conda-forge
qtpy                      1.9.0                      py_0    conda-forge
readline                  8.0                  h7b6447c_0  
retrying                  1.3.3                      py_2    conda-forge
sam-algorithm             0.7.3                    pypi_0    pypi
scanpy                    1.5.1                    pypi_0    pypi
scikit-learn              0.23.1           py36h423224d_0  
scipy                     1.5.0            py36h0b6359f_0  
seaborn                   0.10.1                     py_0  
send2trash                1.5.0                      py_0    conda-forge
setuptools                47.3.1                   py36_0  
setuptools-scm            4.1.2                    pypi_0    pypi
sip                       4.19.8           py36hf484d3e_0  
six                       1.15.0             pyh9f0ad1d_0    conda-forge
snappy                    1.1.8                he6710b0_0  
sqlite                    3.32.3               h62c20be_0  
statsmodels               0.11.1           py36h7b6447c_0  
tbb                       2020.0.133               pypi_0    pypi
terminado                 0.8.3            py36h9f0ad1d_1    conda-forge
testpath                  0.4.4                      py_0    conda-forge
texttable                 1.6.2                    pypi_0    pypi
threadpoolctl             2.1.0              pyh5ca1d4c_0  
tk                        8.6.10               hbc83047_0  
tornado                   6.0.4            py36h8c4c3a4_1    conda-forge
tqdm                      4.46.1                   pypi_0    pypi
traitlets                 4.3.3            py36h9f0ad1d_1    conda-forge
umap-learn                0.4.4                    pypi_0    pypi
wcwidth                   0.2.5              pyh9f0ad1d_0    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.34.2                   py36_0  
widgetsnbextension        3.5.1                    py36_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xz                        5.2.5                h7b6447c_0  
zeromq                    4.3.2                he1b5a44_2    conda-forge
zipp                      3.1.0                      py_0    conda-forge
zlib                      1.2.11               h7b6447c_3  
zstd                      1.4.4                h0b5b093_3 
```

**Comment by giovp** [comment]
this is not related to scanpy, but to sam (scanpy external). Please report the bug in the original repo: https://github.com/atarashansky/self-assembling-manifold
pinging @atarashansky who is possibly most helpful in this.

**Comment by atarashansky** [comment]
Hi @alkhairohr! Strange bug -- never seen it before.

The latest UMAP version I've been using with SAM is `0.4.1`. Can you downgrade UMAP to that version and try again? If that fixes your issue, then I'll add `umap<=0.4.1` requirement to the `setup.py` file for SAM as a stopgap until I figure out the issue.

Meanwhile, I'll try upgrading to `0.4.4` and see if I can reproduce the error.

As @giovp said, because this is a SAM issue, I'll follow up on this issue with a thread on my github repo and ping you there.

Thanks!
